integrations/gitlab/gitlab_integration/events/setup.py

```diff
@@ -0,0 +1,131 @@
+        for token in token_group_override_hooks_mapping:
+            if token not in token_mapping:
+                raise TokenNotFoundException(token)
+            groups_paths.extend(token_group_override_hooks_mapping[token])
+
+        for group_path in groups_paths:
+            if groups_paths.count(group_path) > 1:
+                raise EventListenerConflict(
+                    f"Cannot listen to the same group multiple times. group: {group_path}"
+                )
+            for second_group_path in groups_paths:
+                if second_group_path != group_path and second_group_path.startswith(
+                    group_path
+                ):
+                    raise EventListenerConflict(
+                        "Cannot listen to multiple groups with hierarchy to one another."
+                        f" Group: {second_group_path} is inside group: {group_path}"
+                    )
+
+
+def setup_listeners(gitlab_service: GitlabService, webhook_id: str | int) -> None:
+    handlers = [
+        PushHook(gitlab_service),
+        MergeRequest(gitlab_service),
+        Job(gitlab_service),
+        Issues(gitlab_service),
+        Pipelines(gitlab_service),
+        GroupHook(gitlab_service),
+    ]
+    for handler in handlers:
+        event_ids = [f"{event_name}:{webhook_id}" for event_name in handler.events]
+        event_handler.on(event_ids, handler.on_hook)
+
+
+def setup_system_listeners(gitlab_clients: list[GitlabService]) -> None:
+    handlers: List[Type[HookHandler]] = [
+        PushHook,
+        MergeRequest,
+        Job,
+        Issues,
+        Pipelines,
+        GroupHook,
+    ]
+    for handler in handlers:
+        system_event_handler.on(handler)
+
+    for gitlab_service in gitlab_clients:
+        system_event_handler.add_client(gitlab_service)
+
+
+def setup_application(
+    token_mapping: dict[str, list[str]],
+    gitlab_host: str,
+    app_host: str,
+    use_system_hook: bool,
+    token_group_override_hooks_mapping: dict[str, list[str]],
+) -> None:
+    validate_configuration(token_mapping, token_group_override_hooks_mapping)
+    clients = []
+    for token, group_mapping in token_mapping.items():
+        clients.append(
+            listen_to_webhook_by_token(
+                gitlab_host,
+                app_host,
+                use_system_hook,
+                token,
+                token_group_override_hooks_mapping,
+                group_mapping,
+            )
+        )
+    if use_system_hook:
+        setup_system_listeners(clients)
+
+
+def listen_to_webhook_by_token(
+    gitlab_host: str,
+    app_host: str,
+    use_system_hook: bool,
+    token: str,
+    token_group_override_hooks_mapping: dict[str, list[str]],
+    group_mapping: list[str],
+) -> GitlabService:
+    gitlab_client = Gitlab(gitlab_host, token)
+    gitlab_service = GitlabService(gitlab_client, app_host, group_mapping)
+    if use_system_hook:
+        gitlab_service.create_system_hook()
```

Comment:
this should run only once
pls add a validation to check that if they use system hooks there is only 1 token mapping

Reply:
done